### ISSUE TITLE
Report errors on UAST execution (new column or other solution)

### ISSUE THREAD

**Issue by EgorBu** [src-d/engine#165]
Hi,
Let's improve documentation/examples a bit.
Some behavior of engine is not always clear - example:
```python
from pyspark.sql.functions import col, size
python_files = (engine.repositories.references.head_ref.files.classify_languages()
                .where('lang = "Python"').extract_uasts().cache())
python_files.rdd.map(lambda r: do_something_with_uast(r.uast[0])).count()
```
Code above may fail with `IndexError: list index out of range` because `uast` was not extracted for some files (broken code). It's not obvios that broken code passes `.where('lang = "Python"')`.

I want to create PR with an explanation of this case, please tell me which example is better to update or create a new one.

**Comment by erizocosmico** [comment]
Why is it not obvious? Python detection is performed only by the extension, so no code has actually been checked. UAST is a list and lists always may be empty.

Maybe we can add a method to `UASTsDataFrame` called `get_uast` or something that returns the first uast or `None` if there is nothing. What do you think?

**Comment by EgorBu** [comment]
Not so obvious - several reasons:
1) It's not obvious that broken code passes `.where('lang = "Python"')` - you know that filtering is done by an extension of a file, other users may not know it.
2) Initially, I reported the error to https://github.com/src-d/engine/issues/151 - but it took quite a lot of time to figure out that the reason for mine error is another one - that I explained above.
3) It super hard to debug pyspark errors - there are so many layers of abstraction - so it can be tricky to figure out that there could be some obvious errors.

**Comment by erizocosmico** [comment]
1. We could clarify a bit the part that even if a language has been identified for a file it does not guarantee it will have an UAST. (For example if you have a, let's say, Ruby file. It will be identified but won't have an UAST because there is no Ruby driver)
3. Sadly, nothing we can do about that :stuck_out_tongue:

**Comment by bzz** [comment]
:+1: for better error reporting!

@erizocosmico suggestion of adding a convenience wrapper with documented expectations sounds great.

But @EgorBu its always safer to check for things being empty, instead of assuming non-documented contract, in case of "random data from the internet".

That would help to mitigate 3.

**Comment by ajnavarro** [comment]
We need to research a little bit how to show errors in Rows. I think some spark integrations creates a column with the errors processing the Row if any.

Yep, anyways, as @bzz  said, the code should check for all the possible cases, to avoid errors that stop the processing of the rest of the data unexpectedly.

**Comment by erizocosmico** [comment]
Should we implement the convenience wrapper for `get_uast` then?

**Comment by ajnavarro** [comment]
@erizocosmico having an `Option` type as a column type it's not possible using Spark and DataFrames. We can discuss this functionality on other thread.

**Comment by erizocosmico** [comment]
I don't mean having an `Option`. `uast` is a list of bytearrays. If it's empty it means no uast. So we can do this method:

```python
def get_uast(self, default=None):
    if len(self.uast) == 0:
        return default
    else:
        return self.uast[0]

**Comment by ajnavarro** [comment]
Discused offline with @erizocosmico that we cannot add a method to the Row abstraction, so the `get_uast` method is not possible to implement. Also we shouldn't add functionality outside DataFrame context.

**Comment by mcarmonaa** [comment]
Error reporting is included in this PR #187
